File: src/atcoder/abc135/abc135_f.rb
# Z algorithm
# sとs[i..]の共通接頭辞の長さの配列をO(|s|)で求める
#
# ```
# ZAlgorithm.new("abcbcba").solve # => [7,0,0,0,0,0,1]
# ```
class ZAlgorithm
  attr_reader :n, :s, :ans

  def initialize(s)
    @s = s
    @n = s.size
    @ans = Array.new(n, 0)
    ans[0] = n
  end

  def solve
    i = 1
    j = 0

    while i < n
      while i + j < n && s[j] == s[i + j]
        j += 1
      end

      ans[i] = j

      if j == 0
        i += 1
        next
      end

      k = 1

      while i + k < n && k + ans[k] < j
        ans[i + k] = ans[k]
        k += 1
      end

      i += k
      j -= k
    end

    ans
  end
end

class Graph
  attr_reader :n, :g, :ind

  def initialize(n)
    @n = n
    @g = Array.new(n) { [] }
    @ind = Array.new(n, 0)
  end

  def add(v, nv)
    g[v] << nv
    ind[nv] += 1
  end

  def [](v)
    g[v]
  end

  def tsort
    ans = []
    q = []

    n.times do |v|
      q << v if ind[v].zero?
    end

    while q.size > 0
      v = q.shift
      ans << v

      g[v].each do |nv|
        ind[nv] -= 1
        q << nv if ind[nv].zero?
      end
    end
    ans
  end
end

s = gets.chomp
t = gets.chomp
n = s.size
m = t.size

g = Graph.new(n)

if n < m
  s = s * ((m + n - 1) / n)
end

cnt = ZAlgorithm.new(t + "@" + s + s).solve

n.times do |v|
  i = v + m + 1
  if cnt[i] == m
    nv = (v + m) % n
    g.add v, nv
  end
end

dp = Array.new(n, 0)
ts = g.tsort

if ts.size < n
  pp -1
  exit
end

ts.each do |v|
  g[v].each do |nv|
    dp[nv] = dp[v] + 1
  end
end

pp dp.max

File: src/atcoder/abc135/abc135_f.go
package main

import (
	"container/list"
	"fmt"
	"strings"
)

func ZAlgorithm(s string) []int {
	i := 1
	j := 0
	n := len(s)
	ans := make([]int, n)
	ans[0] = n

	for i < n {
		for i+j < n && s[j] == s[i+j] {
			j++
		}

		ans[i] = j

		if j == 0 {
			i++
			continue
		}

		k := 1

		for i+k < n && k+ans[k] < j {
			ans[i+k] = ans[k]
			k++
		}

		i += k
		j -= k
	}

	return ans
}

type Graph struct {
	n     int
	edges [][]int
	ind   []int
}

func NewGraph(n int) Graph {
	return Graph{
		n,
		make([][]int, n),
		make([]int, n),
	}
}

func (g Graph) Add(v, nv int) {
	g.edges[v] = append(g.edges[v], nv)
	g.ind[nv]++
}

func (g Graph) TSort() []int {
	ans := make([]int, 0)
	q := list.New()

	for v := 0; v < g.n; v++ {
		if g.ind[v] == 0 {
			q.PushBack(v)
		}
	}

	for q.Len() > 0 {
		v := q.Front().Value.(int)
		q.Remove(q.Front())
		ans = append(ans, v)

		for _, nv := range g.edges[v] {
			g.ind[nv]--
			if g.ind[nv] == 0 {
				q.PushBack(nv)
			}
		}
	}

	return ans
}

func main() {
	var s, t string
	fmt.Scan(&s)
	fmt.Scan(&t)

	n := len(s)
	m := len(t)

	if n < m {
		s = strings.Repeat(s, (m+n-1)/n)
	}

	cnt := ZAlgorithm(t + "@" + s + s)

	g := NewGraph(n)

	for v := 0; v < n; v++ {
		i := v + m + 1
		if cnt[i] == m {
			nv := (v + m) % n
			g.Add(v, nv)
		}
	}

	dp := make([]int, n)
	ts := g.TSort()

	if len(ts) < n {
		fmt.Println(-1)
	} else {
		for _, v := range ts {
			for _, nv := range g.edges[v] {
				dp[nv] = dp[v] + 1
			}
		}

		ans := 0
		for _, v := range dp {
			if ans < v {
				ans = v
			}
		}

		fmt.Println(ans)
	}
}
